gossip/bloom.go

```diff
@@ -0,0 +1,92 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"encoding/binary"
+	"hash"
+	"time"
+
+	bloomfilter "github.com/holiman/bloomfilter/v2"
+	"golang.org/x/exp/rand"
+)
+
+var _ hash.Hash64 = (*hasher)(nil)
+
+func NewBloomFilter(m uint64, p float64) (*BloomFilter, error) {
+	bloom, err := bloomfilter.NewOptimal(m, p)
+	if err != nil {
+		return nil, err
+	}
+
+	bloomFilter := &BloomFilter{
+		Bloom: bloom,
+		Salt:  randomSalt(),
+	}
+	return bloomFilter, nil
+}
+
+type BloomFilter struct {
+	Bloom *bloomfilter.Filter
+	Salt  []byte
+}
+
+func (b *BloomFilter) Add(gossipable Gossipable) {
+	salted := hasher{
+		hash: gossipable.GetHash(),
+		salt: b.Salt,
+	}
+	b.Bloom.Add(salted)
+}
+
+func (b *BloomFilter) Has(gossipable Gossipable) bool {
+	salted := hasher{
+		hash: gossipable.GetHash(),
+		salt: b.Salt,
+	}
+	return b.Bloom.Contains(salted)
+}
+
+// ResetBloomFilterIfNeeded resets a bloom filter if it breaches a ratio of
+// filled elements. Returns true if the bloom filter was reset.
+func ResetBloomFilterIfNeeded(
```

Comment:
I think this will be something we end up needing for other components that consume this package, eventually someone will need to reset their bloom filter as over time the amount of collisions will increase.